flips/2022-03-23-contract-imports-syntax.md

```diff
@@ -0,0 +1,150 @@
+
+Example:
+`import Foo from ./Foo.cdc`
+
+**Mechanics**
+
+Source code is loaded and parsed with the Cadence library, the import statements are iterated over and replaced with the values set in `flow.json` configuration. 
+
+**Used by**
+
+- VSCode Extension
+- CLI
+
+### Current Implementation Problems
+
+**Imports referencing a relative path**
+
+- there will not always be a contract present locally (or at least it shouldn't be as the network is the source of truth for a contract and downloading the contract has definite downsides especially contracts that are still able to update)
+- potential problems across different file systems
+
+**Imports referencing a placeholder**
+
+- Not a valid cadence syntax
+- Impossible to know what contract is `0xFoo` really referencing by reading the contract code, the resolution of contract location is hidden in the complexity of the code
+
+### The Current State
+
+We can observe some common patterns established in the community:
+
+- Contracts are published on version control platforms (eg Github) and the best-established practice is for the code to be published on its own without other application code (which usually is part of another repository)
+- The community is using and switching between CLI and SDKs (FCL) and is facing problems between different import formats. This problem is validated with the existence of tools that handle switching between import formats (such as [github link])
+
+### Possible Solutions?
+
+**New Import Syntax**
+
+`import <declarations to import> from "<namespaced location>"`
+
+Example:
+
+`import NonFungibleToken from "onflow/NonFungibleToken"`
+
+The source files should be accompanied by a configuration/dependency resolution file, defining locations of contracts addresses on a specific network. Including this file is beneficial so developers exploring contract source files in public repositories can understand which contracts this Cadence code imports. Think of this as go.mod, package.json or any other implementation of this idea where source code that imports something needs to define what that is in the supporting "configuration" file.
+
+Example flow.json for the above syntax:
+
+```jsx
+{
+...
+"contracts": {
+	"onflow/NonFungibleToken": "./NonFungibleToken.cdc",
+	"onflow/FungibleToken": {
+		"testnet": "github.com/flow/FungibleToken/Fungible.cdc",
+		"mainnet": "0x2",
+		"emulator": "./FungibleToken.cdc", 
+		"e2e-testnet": "0x3"
+	}
+}
+...
+}
+```
+
+Another benefit to this file is to have a way to define mock contracts for testing. You wouldn't be limited to defining one address per network but you could define multiple environments.
+
+The above example is just an idea for values this file needs to include, the file format is up for discussion, but we could argue `flow.json` might be a good place to put this information especially since it can be used that it just defines this section and nothing else.
+
+### Implementation
```

Comment:
Maybe note here that there is no implementation needed in Cadence itself, as it already supports string locations (i.e. the syntax `import "contract"`)

Reply:
good flag thanks